Unifying all troubleshooting guides so far (#17553)
metabase/metabase: docs/troubleshooting-guide/my-dashboard-is-slow.md

> # My dashboard is slow

Reviewer: Consider adding links to the [Making dashboards faster](https://www.metabase.com/learn/administration/making-dashboards-faster) and [Metabase at scale](https://www.metabase.com/learn/administration/metabase-at-scale.html) in this opening section (in addition to the links below).

Author: added

2. Look at the size of the result set for each question that is slow. If it contains hundreds of columns or many thousands of rows, see if you can modify the question to return a smaller result set.
3. Run the same SQL that Metabase is running but using a different tool. For example, if you have created a question using the Notebook Editor and you are using Postgres as your database, you can view the SQL, copy it, and run it from the command line using `psql`. (If you have written the question in SQL you can just copy and paste it.) If the query runs noticeably faster this way than it does when you run it through Metabase, the problem is almost certainly one of the ones described above.
4. Run the query with a JDBC-based tool like [DBeaver][dbeaver], which is also Java-based and uses most of the same database drivers as Metabase. Using this, and running the test query from the same machine that Metabase is running on, can help you determine if the problem is a network bottleneck.
5. If the query runs slowly when sent by another tool, see if you can write a SQL query that calculates the same result as the question you have built in Metabase, but does so more quickly.

> ## Are we generating inefficient SQL?

Reviewer: ```suggestion
## Is Metabase generating inefficient SQL?
```

> **Root cause:** We save questions created graphically in Metabase Query Language (MBQL), then translate MBQL into queries for particular back-end databases. We create the most efficient queries we can, but for the sake of portability, we don't take advantage of every database's idiosyncracies, so sometimes a GUI question will be slower than the equivalent hand-written SQL.

Reviewer: ```suggestion
**Root cause:** Metabase saves questions created graphically in Metabase Query Language (MBQL), then translate MBQL into queries for particular back-end databases. It creates the most efficient queries it can, but for the sake of portability, it doesn't take advantage of every database's idiosyncracies, so sometimes a GUI question will be slower than the equivalent hand-written SQL.
```

**Steps to take:**

1. Check if you have the most recent version of Metabase: we fix problems as they're reported, and updating Metabase may make your problem go away.
2. You can use your SQL in place of the code we generate, and [make its result available][organizing-sql] to people who prefer the Notebook Editor as a starting point for their questions.

> 3. Please file a bug report to help us figure out how to generate better SQL.

Reviewer: Link to https://www.metabase.com/docs/latest/troubleshooting-guide/bugs.html?

Author: added

## Do you have the right database schema?

> **Root cause:** If the database schema is not designed well, questions cannot run quickly.

Reviewer: ```suggestion
**Root cause:** If the database schema is poorly designed, questions cannot run quickly.
```

**Steps to take:** The difference between OLTP and OLAP, and how to design database schemas to support them, are out of scope for this troubleshooting guide. In brief, you must look at the data schema of the data warehouse, or ask the database administrator whether the database is designed for online transaction processing (OLTP) or online analytical processing (OLAP). Metabase is an OLAP application; if the database schema is designed for OLTP, you may need to create views that reorganize the data.

Similarly, you probably don't need indexes for simple tables with a few tens of thousands of rows, but you almost certainly *do* if you have a few million rows. All of this is very dependent on the underlying database: Redshift can easily handle millions of rows with thousands of columns, but MySQL or PostgreSQL may require a star schema designed for OLAP to deliver the performance you need.

> ## Is the answer to your question not being cached?

Reviewer: ```suggestion
## Is Metabase not caching the answer to your question?
```

**Root cause:** By default caching is disabled so that we always re-run every question. However, if your data is only being updated every few seconds or minutes, you will improve performance by enabling caching. Note that:

1. Ad hoc queries and exports are *not* cached, so doing a lot of either can impact performance.
2. Since cached values are stored in the application database they will still be there if Metabase restarts.
3. Each question (and any filter combination) is its own query, so if different users are viewing the same question with different filters, each will have to load once before it's cached. This is particularly obvious with [data sandboxing][data-sandboxing]: filtering the data based on the user's identity means that each user's question is slightly different.

**Steps to take:**

> 1. Go to Admin Panel > Settings > Caching to see whether caching is enabled or not.

Reviewer: ```suggestion
1. Go to Admin Panel > Settings > Caching to see if caching is enabled.
```
